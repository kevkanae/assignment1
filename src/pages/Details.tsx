import {
  Box,
  Button,
  Card,
  CardActions,
  CardContent,
  Typography,
} from "@mui/material";
import { useQuery } from "react-query";
import { useParams } from "react-router-dom";
import CountryCard from "../components/CountryCard";
import { ICountryDetails } from "../interfaces/CountryDetails";
import { getCountryDetailsAPI } from "../services/GetCountryDetails";
import { flexCol } from "../styles/flex";

const Details = () => {
  let params = useParams();
  const { data, isFetched, isError, error } = useQuery<
    ICountryDetails[],
    Error
  >(["product", params.countryName], () =>
    getCountryDetailsAPI(params.countryName as string)
  );

  if (isError) {
    return (
      <>
        <Box>There was an Error: {error}</Box>
      </>
    );
  } else if (isFetched && !isError && data) {
    return (
      <Box
        sx={{
          ...flexCol,
          justifyContent: "center",
          height: "100vh",
          width: "100vw",
        }}
      >
        <CountryCard data={data[0] as ICountryDetails} />
      </Box>
    );
  } else {
    return <Box>Loading...</Box>;
  }
};

export default Details;
