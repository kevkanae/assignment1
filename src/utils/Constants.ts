export const countryDetailsURL = "https://restcountries.com/v3.1/name/";

const key = "74964c7ef563f1dded7f8efdf8df8620";
export const weatherDetailsURL = `http://api.weatherstack.com/current?access_key=${key}&query=`;
