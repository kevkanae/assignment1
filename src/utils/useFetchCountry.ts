import { useQuery } from "react-query";
import { ICountryDetails } from "../interfaces/CountryDetails";
import { getCountryDetailsAPI } from "../Services/GetCountryDetails";

export const useFetchCountry = (country: string) => {
  const { data, isFetched, isError, error, isSuccess } = useQuery<
    ICountryDetails[],
    Error
  >(["product", country], () => getCountryDetailsAPI(country as string));

  return { data, error, isFetched, isError, isSuccess };
};
