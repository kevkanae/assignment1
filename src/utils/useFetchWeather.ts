import { useQuery } from "react-query";
import { IWeatherDetails } from "../interfaces/WeatherDetails";
import { getWeatherDetailsAPI } from "../services/GetWeatherDetails";

export const useFetchWeather = (capital: string) => {
  const { data, isFetched, isError, error, isSuccess } = useQuery<
    IWeatherDetails,
    Error
  >(["product", capital], () => getWeatherDetailsAPI(capital));

  return { data, error, isFetched, isError, isSuccess };
};
