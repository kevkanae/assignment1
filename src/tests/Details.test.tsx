import { render, screen } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";
import { QueryClient, QueryClientProvider } from "react-query";
import { MemoryRouter, BrowserRouter } from "react-router-dom";
import Details from "../pages/Details";
import renderer from "react-test-renderer";
import { useFetchCountry } from "../utils/useFetchCountry";
const queryClient = new QueryClient();

test("Snapshot of Home component", () => {
  const comp = renderer.create(
    <BrowserRouter>
      <QueryClientProvider client={queryClient} contextSharing={true}>
        <Details />
      </QueryClientProvider>
    </BrowserRouter>
  );
  let tree = comp.toJSON();
  expect(tree).toMatchSnapshot();
});

describe("does the url params", () => {
  test("exist", () => {
    render(
      <MemoryRouter initialEntries={["/:countryName"]}>
        <QueryClientProvider client={queryClient} contextSharing={true}>
          <Details />
        </QueryClientProvider>
      </MemoryRouter>
    );

    jest.mock("react-router-dom", () => ({
      ...jest.requireActual("react-router-dom"),
      useParams: () => ({
        countryName: "Japan",
      }),
    }));
  });
});

describe("does react query", () => {
  const createWrapper = () => {
    const queryClient = new QueryClient();
    return ({ children }: any) => (
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    );
  };

  test("work", async () => {
    const { result, waitFor }: any = renderHook(
      () => useFetchCountry("Japan"),
      {
        wrapper: createWrapper(),
      }
    );
    await waitFor(() => result.current.isSuccess, { timeout: 8000 });

    expect(result.current.data).toBeDefined();
  }, 10000);

  test("is the button present", () => {
    render(
      <BrowserRouter>
        <QueryClientProvider client={queryClient} contextSharing={true}>
          <Details />
        </QueryClientProvider>
      </BrowserRouter>
    );
    const btn = screen.getByRole("loadbtn");
    expect(btn).toBeInTheDocument();
  });
});
