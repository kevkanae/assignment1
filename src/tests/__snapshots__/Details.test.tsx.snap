// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of Home component 1`] = `
<div
  className="MuiBox-root css-q7q1yu"
  role="loadbtn"
>
  <span
    className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1gw1ctm-MuiCircularProgress-root"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </span>
</div>
`;
